File: dbflow/src/main/java/com/raizlabs/android/dbflow/config/FlowConfig.java
package com.raizlabs.android.dbflow.config;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

/**
 * Description: The main configuration instance for DBFlow. This
 */
public final class FlowConfig {

    public static FlowConfig.Builder builder(Context context) {
        return new FlowConfig.Builder(context);
    }

    private final Set<Class<? extends DatabaseHolder>> databaseHolders;
    private final Map<Class<?>, DatabaseConfig> databaseConfigMap;
    private final Context context;
    private final boolean openDatabasesOnInit;

    FlowConfig(Builder builder) {
        databaseHolders = Collections.unmodifiableSet(builder.databaseHolders);
        databaseConfigMap = builder.databaseConfigMap;
        context = builder.context;
        openDatabasesOnInit = builder.openDatabasesOnInit;
    }

    @NonNull
    public Set<Class<? extends DatabaseHolder>> databaseHolders() {
        return databaseHolders;
    }

    @NonNull
    public Map<Class<?>, DatabaseConfig> databaseConfigMap() {
        return databaseConfigMap;
    }

    @Nullable
    public DatabaseConfig getConfigForDatabase(@NonNull Class<?> databaseClass) {
        return databaseConfigMap().get(databaseClass);
    }

    @NonNull
    public Context getContext() {
        return context;
    }

    public boolean openDatabasesOnInit() {
        return openDatabasesOnInit;
    }

    public static class Builder {

        final Context context;
        Set<Class<? extends DatabaseHolder>> databaseHolders = new HashSet<>();
        final Map<Class<?>, DatabaseConfig> databaseConfigMap = new HashMap<>();
        boolean openDatabasesOnInit;

        public Builder(Context context) {
            this.context = context.getApplicationContext();
        }

        @NonNull
        public Builder addDatabaseHolder(@NonNull Class<? extends DatabaseHolder> databaseHolderClass) {
            databaseHolders.add(databaseHolderClass);
            return this;
        }

        @NonNull
        public Builder addDatabaseConfig(@NonNull DatabaseConfig databaseConfig) {
            databaseConfigMap.put(databaseConfig.databaseClass(), databaseConfig);
            return this;
        }

        /**
         * @param openDatabasesOnInit true if we want all databases open.
         * @return True to open all associated databases in DBFlow on calling of {@link FlowManager#init(FlowConfig)}
         */
        @NonNull
        public Builder openDatabasesOnInit(boolean openDatabasesOnInit) {
            this.openDatabasesOnInit = openDatabasesOnInit;
            return this;
        }

        @NonNull
        public FlowConfig build() {
            return new FlowConfig(this);
        }
    }
}

File: dbflow/src/main/java/com/raizlabs/android/dbflow/config/FlowConfig.kt
package com.raizlabs.android.dbflow.config

import android.content.Context
import java.util.*

/**
 * Description: The main configuration instance for DBFlow. This
 */
class FlowConfig(val context: Context,
                 val databaseHolders: Set<Class<out DatabaseHolder>> = setOf(),
                 val databaseConfigMap: Map<Class<*>, DatabaseConfig> = mapOf(),
                 val openDatabasesOnInit: Boolean = false) {

    internal constructor(builder: Builder) : this(
            databaseHolders = Collections.unmodifiableSet(builder.databaseHolders),
            databaseConfigMap = builder.databaseConfigMap,
            context = builder.context,
            openDatabasesOnInit = builder.openDatabasesOnInit
    )

    fun getConfigForDatabase(databaseClass: Class<*>): DatabaseConfig? {
        return databaseConfigMap[databaseClass]
    }

    class Builder(context: Context) {

        internal val context: Context = context.applicationContext
        internal var databaseHolders: MutableSet<Class<out DatabaseHolder>> = HashSet()
        internal val databaseConfigMap: MutableMap<Class<*>, DatabaseConfig> = HashMap()
        internal var openDatabasesOnInit: Boolean = false

        fun addDatabaseHolder(databaseHolderClass: Class<out DatabaseHolder>) = apply {
            databaseHolders.add(databaseHolderClass)
        }

        fun addDatabaseConfig(databaseConfig: DatabaseConfig) = apply {
            databaseConfigMap.put(databaseConfig.databaseClass, databaseConfig)
        }

        /**
         * @param openDatabasesOnInit true if we want all databases open.
         * @return True to open all associated databases in DBFlow on calling of [FlowManager.init]
         */
        fun openDatabasesOnInit(openDatabasesOnInit: Boolean) = apply {
            this.openDatabasesOnInit = openDatabasesOnInit
        }

        fun build() = FlowConfig(this)
    }

    companion object {

        fun builder(context: Context): FlowConfig.Builder {
            return FlowConfig.Builder(context)
        }
    }
}
